kimchi/src/snarky/boolean.rs

```diff
@@ -0,0 +1,192 @@
+        Self(cvars[0].clone())
+    }
+
+    fn check(&self, cs: &mut RunState<F>) {
+        // TODO: annotation?
+        cs.assert_(None, vec![BasicSnarkyConstraint::Boolean(self.0.clone())]);
+    }
+
+    fn deserialize(&self) -> (Self::OutOfCircuit, Self::Auxiliary) {
+        todo!()
+    }
+
+    fn serialize(out_of_circuit: Self::OutOfCircuit, aux: Self::Auxiliary) -> Self {
+        todo!()
+    }
+
+    fn constraint_system_auxiliary() -> Self::Auxiliary {
+        todo!()
+    }
+
+    fn value_to_field_elements(x: &Self::OutOfCircuit) -> (Vec<F>, Self::Auxiliary) {
+        todo!()
+    }
+
+    fn value_of_field_elements(x: (Vec<F>, Self::Auxiliary)) -> Self::OutOfCircuit {
+        todo!()
+    }
+}
+
+impl<F> Boolean<F>
+where
+    F: PrimeField,
+{
+    pub fn true_() -> Self {
+        Self(CVar::Constant(F::one()))
+    }
+
+    pub fn false_() -> Self {
+        Self(CVar::Constant(F::zero()))
+    }
+
+    pub fn not(&self) -> Self {
+        Self(Self::true_().0 - &self.0)
+    }
+
+    pub fn if_(&self, then_: Self, else_: Self) -> Self {
+        todo!()
+    }
+
+    pub fn and(&self, other: &Self, cs: &mut RunState<F>) -> Self {
+        Self(self.0.mul(&other.0, Some("bool.and"), cs))
+    }
+
+    pub fn or(&self, other: &Self, cs: &mut RunState<F>) -> Self {
+        let both_false = self.not().and(&other.not(), cs);
+        both_false.not()
+    }
+
+    pub fn any(xs: &[&Self], cs: &mut RunState<F>) -> Self {
+        if xs.is_empty() {
+            return Self::false_(); // TODO: shouldn't we panic instead?
+        } else if xs.len() == 1 {
+            return xs[0].clone();
+        } else if xs.len() == 2 {
+            return xs[0].or(&xs[1], cs); // TODO: is this better than below?
+        }
+
+        let zero = CVar::Constant(F::zero());
+
+        let xs: Vec<_> = xs.into_iter().map(|x| &x.0).collect();
+        let sum = CVar::sum(&xs);
+        let all_zero = sum.equal(cs, &zero);
+
+        all_zero.not()
+    }
+
+    pub fn all(xs: &[Self], cs: &mut RunState<F>) -> Self {
+        if xs.is_empty() {
+            return Self::true_(); // TODO: shouldn't we panic instead?
+        } else if xs.len() == 1 {
+            return xs[0].clone();
+        } else if xs.len() == 2 {
+            return xs[0].and(&xs[1], cs); // TODO: is this better than below?
+        }
+
+        let expected = CVar::Constant(F::from(xs.len() as u64));
+        let xs: Vec<_> = xs.iter().map(|x| &x.0).collect();
+        let sum = CVar::sum(&xs);
+
+        sum.equal(cs, &expected)
+    }
+
+    pub fn to_constant(&self) -> Option<bool> {
+        match self.0 {
+            CVar::Constant(x) => Some(x == F::one()),
+            _ => None,
+        }
+    }
+
+    pub fn xor(&self, other: &Self, state: &mut RunState<F>) -> Self {
+        match (self.to_constant(), other.to_constant()) {
+            (Some(x), Some(y)) => {
+                // (var_of_value (Caml.not (Bool.equal b1 b2)))
+                todo!()
```

Comment:
You can remove this case and let it fall through to `(Some(true), _)` or `(Some(false), _)`, if that makes it easier.